### PR TITLE
CUMULUS-2192: Update parallel write logic for executions

### PR DESCRIPTION
**Pull request by markdboyd**
**Summary:** Summary of changes

Addresses [CUMULUS-2192: Update all execution database writes to write to both Dynamo and RDS](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2192)

## Changes

- Update version of Knex to version 0.12.12 include new upsert functionality: https://github.com/knex/knex/pull/3763
- Update execution writes in sfEventSqsToDbRecords Lambda to use upsert logic
- Add unit tests for execution upsert behavior

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [ ] Adhoc testing

### CODE REVIEW

File: packages/api/lambdas/sf-event-sqs-to-db-records/index.js
Comment (markdboyd) [inline packages/api/lambdas/sf-event-sqs-to-db-records/index.js:69]: This was refactored because I need the cumulus ID for the async operation or parent execution (if they exist) to be passed to the execution write logic on lines 91-93

File: packages/api/lambdas/sf-event-sqs-to-db-records/utils.js
Comment (markdboyd) [inline packages/api/lambdas/sf-event-sqs-to-db-records/utils.js:40]: As mentioned above, I realized that I needed these helpers to return the cumulus ID, if the record exists, not just to tell me whether the record exists or not

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js
Comment (markdboyd) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js:37]: A bonus of the refactor is that this method is a lot simpler now

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js
Comment (markdboyd) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js:90]: It rocks so hard that Knex implemented this

File: packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-utils.js
Comment (markdboyd) [inline packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-utils.js:13]: A very nice side effect of the refactoring is that there is less need for mocking, which is 💯

File: CHANGELOG.md
Comment (Jkovarik) [inline CHANGELOG.md:62]: :100:

File: packages/api/lambdas/sf-event-sqs-to-db-records/index.js
Comment (Jkovarik) [reply packages/api/lambdas/sf-event-sqs-to-db-records/index.js:69]: This seems :100: code wise.     I'm wondering if there's a more optimized way of getting these connection/query wise, but I suppose for the moment that's YAGNI.   Something possibly to stash when we get to performance tuning though.

File: packages/api/lambdas/sf-event-sqs-to-db-records/utils.js
Comment (Jkovarik) [reply packages/api/lambdas/sf-event-sqs-to-db-records/utils.js:40]: 👍   Also if I didn't mention it before, nice job moving the selection logic out to @cumulus/message

File: packages/api/lambdas/sf-event-sqs-to-db-records/utils.js
Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records/utils.js:46]: Nit: I think this came up in a prior PR, it still feels odd having this logic in the *api* package, but perhaps that's more that this *lambda* being there feels odd.    Not something to fix for this epic/PR, but bears consideration later if the lambda and/or these utils should be re-orged a bit.

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js
Comment (Jkovarik) [reply packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js:37]: Probably dumb question - are collections currently required for Executions?    Didn't we make that optional so non-data oriented executions could occur?    The collection doesn't appear to be required in the API schema, and the FK isn't set to be non-nullable on the Postgres migration. 

  In this code if the call further up the stack in the lambda index.js getMessageCollectionCumulusId(cumulusMessage, knex) can return `undefined` and isNil(undefined) resolves to true. 

Same for async. 

The original call to hasNoAsyncOpOrExists would return true if there was no async op ID defined in the message *or* if there was one in the message and it was in the DB. 

That doesn't appear to be true anymore: 


```
const getMessageAsyncOperationCumulusId = async (cumulusMessage, knex) => {
  try {
    const asyncOperationId = getMessageAsyncOperationId(cumulusMessage);
    if (!asyncOperationId) {
      throw new Error('Could not find async operation in message');
    }
    const asyncOperation = await knex(tableNames.asyncOperations).where({
      id: asyncOperationId,
    }).first();
    if (!isRecordDefined(asyncOperation)) {
      throw new Error(`Could not find async operation with id ${asyncOperationId}`);
    }
    return asyncOperation.cumulus_id;
  } catch (error) {
    log.error(error);
    return undefined;
  }
};

```

I'm sure I'm missing/not thinking about something right here.... 🤔

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js
Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js:42]: Nit: I'd feel a lot better if this was typescripted and used the Execution typings.     I also want a pony, so... totally admit this isn't scoped for this PR.  See comments above re: the organization/location of this lambda.

File: packages/message/src/Executions.ts
Comment (Jkovarik) [inline packages/message/src/Executions.ts:18]: 👍

**Review by Jkovarik — CHANGES_REQUESTED** [review]
Review still in progress, dumping pending comments so you can review the ones I've already added.

File: packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js
Comment (Jkovarik) [inline packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js:5]: :tada:

File: packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js
Comment (Jkovarik) [inline packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js:124]: See above concern re: write logic - we need changes/tests around async/collection being not present in the message.

File: packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js
Comment (Jkovarik) [inline packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js:307]: Looking at what's being evaluated here.... we're really testing: 

1) That the method under test passes the FKs to buildExecutionRecord (the actual building of the record should be tested in that methods tests)... which it should do each time. 
2) That knex can insert whatever comes back from #1 and returns the expected ID, and in this instance in particular it includes *a specific FK* 

Is there a reason we shouldn't make all of the FK relationships a setup for each write test below and just remove this test entirely (possibly more rad) , or alternately mock out the call to build record and test that it's called with the expected values (perhaps less rad)?      I'm not sure there's a valid scenario these tests cover where the reverse case (no FKs) fails the knex insert where a FK constrained object *would*. 

🤔

File: packages/message/src/Executions.ts
Comment (Jkovarik) [inline packages/message/src/Executions.ts:206]: Minor nit: should these be in something like workflow.ts?

**Review by Jkovarik — CHANGES_REQUESTED** [review]
Just a few more comments.   I didn't look too closely at the tests related to the requested changes, I'll re-review once that's sorted.     Aside the write concern, this all looks really good!

File: packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js
Comment (Jkovarik) [inline packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js:334]: This should be removed on resolution of the discussion at L307

File: packages/message/src/Executions.ts
Comment (botanical) [inline packages/message/src/Executions.ts:163]: This is a nit but that comment confuses me. Did you mean `Get the workflow tasks'?

Comment (markdboyd) [reply packages/message/src/Executions.ts:163]: I meant to say "Get the workflow tasks in a workflow message". Would that be better?

Comment (botanical) [reply packages/message/src/Executions.ts:163]: Yes, that works

Comment (markdboyd) [reply packages/message/src/Executions.ts]: 038e5edd614f837d0ef0c20e25490ce998457d03

File: packages/message/src/Executions.ts
Comment (markdboyd) [reply packages/message/src/Executions.ts]: 149161f936f50fee79f1ab78d061559cc950e453

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js
Comment (markdboyd) [reply packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js]: Your concerns were very valid. The logic has been refactored to account for the fact that we need to know both whether there is a reference to an object in the message **and** whether that object exists in the Postgres database.

File: packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js
Comment (markdboyd) [reply packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js]: With the factor, see the new tests:

https://github.com/nasa/cumulus/pull/2005/files?file-filters%5B%5D=.js#diff-74fdb6cc3f08eb88163d987562f8fd73a7181eba22a2b71c6b2490d5d10884a8R99
https://github.com/nasa/cumulus/pull/2005/files?file-filters%5B%5D=.js#diff-74fdb6cc3f08eb88163d987562f8fd73a7181eba22a2b71c6b2490d5d10884a8R112
https://github.com/nasa/cumulus/pull/2005/files?file-filters%5B%5D=.js#diff-74fdb6cc3f08eb88163d987562f8fd73a7181eba22a2b71c6b2490d5d10884a8R125

File: packages/api/models/granules.js
Comment (markdboyd) [inline packages/api/models/granules.js:66]: This is solely to allow for mocking from the unit tests to avoid unnecessary error logs, though I think it's a nice improvement regardless

File: packages/api/tests/models/granules/test-granule-generateGranuleRecord.js
Comment (markdboyd) [inline packages/api/tests/models/granules/test-granule-generateGranuleRecord.js:117]: This is all side effects of a refactor to make mocking easier

File: packages/api/lambdas/sf-event-sqs-to-db-records/index.js
Comment (botanical) [inline packages/api/lambdas/sf-event-sqs-to-db-records/index.js:145]: Why is awaiting no longer needed for `writeGranules`?

Comment (markdboyd) [reply packages/api/lambdas/sf-event-sqs-to-db-records/index.js:145]: Very good question. Originally, `writeGranules` the final statement in a `try` block and its value was supposed to be returned if it did not throw an error. The "return await" is really necessary for the second part - to ensure any error from the function is thrown into the `catch` block, otherwise you're just returning the un-evaluated promise of `writeGranules` so any error from it will only be thrown on a call to the parent function, `writeRecords`.  

But since `writeGranules` is no longer in a try/catch block (because I was just re-throwing the error so the try/catch wasn't really providing any value), I can just `return writeGranules` so that `await writeRecords` effectively awaits the promise returned by `writeGranules`. 

Jake Archibald explains it better than I would: https://jakearchibald.com/2017/await-vs-return-vs-return-await/

File: packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js
Comment (markdboyd) [reply packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js]: I settled on this: 9e627ab56810417dc113c9394c0cd0128c2e4f78

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js
Comment (botanical) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js:32]:
```
noMessageCollectionOrExists
noMessageAsyncOperationOrExists
noMessageParentExecutionOrExists
```
It wasn't initially clear to me why these were named this way until you explained it. I understood the check for a null `messageCollectionNameVersion` and so on, but not the not null check. I think, for clarity, adding `InPostgres` to the variable names would make that more explicit.

```suggestion
  const noMessageCollectionOrExistsInPostgres = isNil(messageCollectionNameVersion)
    || !isNil(collectionCumulusId);
```

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js
Comment (botanical) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js:34]:
```suggestion
  const noMessageAsyncOperationOrExistsInPostgres = isNil(messageAsyncOperationId)
```

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js
Comment (botanical) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js:36]:
```suggestion
  const noMessageParentExecutionOrExistsInPostgres = isNil(messageParentExecutionArn)
```

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js
Comment (botanical) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js:39]:
```suggestion
  return noMessageCollectionOrExistsInPostgres
    && noMessageAsyncOperationOrExistsInPostgres
    && noMessageParentExecutionOrExistsInPostgres;
```

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js
Comment (markdboyd) [reply packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js]: Yeah, good feedback and suggestions. Made those changes here: 870aee35857b8f43b284a5cb57950e1c2895492d

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js
Comment (markdboyd) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js:32]: @Jkovarik (or whoever) - does this logic make sense? it's basically intending to say, "either there **is no** collection on the message or there **is** a collection in Postgres". Something about using the `||` makes me worry it's not safe, but there are unit tests here

Comment (Jkovarik) [reply packages/api/lambdas/sf-event-sqs-to-db-records/write-execution.js:32]: isNil(A) || !isNil(B) 

Works.   Something like "messageAsyncReferencesCorrect" for the variables might work better, though.

**Review by Jkovarik — APPROVED** [review]
I have a few more comments, but nothing to hold up the PR.   Approving, pending discussion re: the upsert concern @markdboyd  had on Friday.

File: packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-index.js
Comment (markdboyd) [reply packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-index.js]: 0e1e035755fc4a8903ab034d03dd863275184b18

**Pull request merged**
